docs/architecture/adr-044-lite-client-with-weak-subjectivity.md

```diff
@@ -121,6 +124,176 @@ network usage.
+
+```go
+type mode int
+
+const (
+	sequential mode = iota
+	bisecting
+)
+
+// default mode - DefaultBisectingVerification
+type Verifier struct {
+	chainID            string
+	options            TrustOptions
+	lastVerifiedHeight int64
+	logger             log.Logger
+
+	mode       mode
+	trustLevel float
+
+	// Source of new headers.
+	source Provider
+
+	// Alternative sources for checking the primary for misbehavior by comparing data.
+	// If the primary misbehaves, we report the evidence to them.
+	verifiers []ProviderAndEvidenceReporter
+
+	// Where trusted headers are stored.
+	trusted PersistentProvider
+}
+```
+
+Since providers themselves don't know when they have received a new header (or
+may choose to do so upon a request), we must add a new function to `Verifier` -
+`Verify(height int64) error` (0 - latest). It will try to fetch a new header &
+validator set and verify it. nop if already verified.
+
+**Sequential vs bisecting verifier**
+
+Verifier should use bisection by default, but provide options to choose a
+different mode OR tweak bisection.
+
+```go
+func SequentialVerification() Option {
+	return func(v *Verifier) {
+		v.mode = sequential
+	}
+}
+
+// trustLevel - maximum change between two not consequitive headers in terms of
+// validators & their respective voting power, required to trust a new header
+// (default: 1/3).
+func BisectingVerification(trustLevel float) Option {
+	if trustLevel > 1 || trustLevel < 1/3 {
+		panic(fmt.Sprintf("trustLevel must be within [1/3, 1], given %v", trustLevel))
+	}
+
+	return func(v *Verifier) {
+		v.mode = bisecting
+		v.trustLevel = trustLevel
+	}
+}
+
+var DefaultBisectingVerification = func() Option {
+	return BisectingVerification(1/3)
+}
+```
+
+Once we verified the header, we will need to store it somewhere.
+
+```
+type PersistentProvider interface {
+	Provider
+
+	SaveFullCommit(fc FullCommit) error
+}
+```
+
+In case of the proxy it will be a `db` provider (levelDB + in-memory cache in
+front). For IBC, it will be a `keeper` provider.
+
+**Minimal test for (1)**
+
+```go
+c, err := lite.NewClient(
+	chainID,
+	lite.TrustOptions{TrustPeriod: 336 * time.Hour},
+	rpcclient.NewHTTP(remote1, "/websocket"),
+)
+require.NoError(t, err)
+
+commit, err := c.Commit()
+require.NoError(t, err)
+assert.Equal(t, chainID, commit.ChainID)
+```
+
+`lite.Client` here is a `struct`, which uses `Verifier` and exposes
+`rpcclient.Client` API.
+
+```go
+type Client struct {
+	verifier *Verifier
+	client   rpcclient.Client
+}
+
+var rpcclient.Client = (*Client)(nil)
+```
+
+**Minimal test for (2)**
+
+```go
+c, err := lite.NewVerifier(
+	chainID,
+	lite.TrustOptions{TrustPeriod: 24 * time.Hour},
+	ibc.New(chainID),
+	Trusted(ibcKeeper{}),
+)
+require.NoError(t, err)
+
+err = c.Verify(height)
+require.NoError(t, err)
+```
+
+**Evidence Handling and Reporting**
+
+light client should also be able to submit evidence of malfeasance and handle
+evidence coming from a full node or another source.
```

Comment:
Why lite client need to handle evidence coming from a full node? What would be the use case?